tools/install-monorepo.sh

```diff
@@ -15,143 +17,344 @@ function usage {
+
+		# Something's wrong, so abort.
+		if ! has_command nvm; then
+			die "Unable to install nvm!"
 		fi
+		echo "Installed nvm."
 	fi
-fi
 
-source .github/versions.sh
-info "Installing and linking PHP $PHP_VERSION"
-brew install php@$PHP_VERSION
-brew link php@$PHP_VERSION
+	echo "nvm: available"
+}
+
+# Checks for and installs node.
+function do_node {
+	good_node_version=$(cat .nvmrc)
+	if ! has_command node; then
+		echo "node: not found"
+		echo "Installing node..."
+		nvm install
+		# Something's wrong, so abort.
+		if ! has_command node; then
+			die "Unable to install node!"
+		fi
+		echo "Node installed."
+	elif [[ $(node --version) != v"$good_node_version" ]]; then
+		# Install correct version if not yet installed
+		if ! nvm list node | grep -q "$good_node_version"; then
+			echo "node: needed version not installed"
+			echo "Installing correct node version..."
+			nvm install
+			# Something's wrong, so abort.
+			if ! nvm list node | grep -q "$good_node_version"; then
+				die "Unable to needed version of node!"
+			fi
+			echo "Correct version of node installed."
+		fi
+	fi
+	# Switch node version
+	nvm use &> /dev/null
+	echo "node: available"
+}
 
-info "Checking composer..."
-if ! command -v composer &>/dev/null; then
-	info "Installing Composer"
-	EXPECTED_CHECKSUM="$(php -r 'copy("https://composer.github.io/installer.sig", "php://stdout");')"
-	php -r "copy('https://getcomposer.org/installer', 'composer-setup.php');"
-	ACTUAL_CHECKSUM="$(php -r "echo hash_file('sha384', 'composer-setup.php');")"
+function do_bash {
+	if [[ -z "${BASH_VERSINFO}" || -z "${BASH_VERSINFO[0]}" || ${BASH_VERSINFO[0]} -lt 4 ]]; then
+		echo "bash 4+: not found"
+		echo "Installing bash..."
+		brew install bash
+		if [[ -z "${BASH_VERSINFO}" || -z "${BASH_VERSINFO[0]}" || ${BASH_VERSINFO[0]} -lt 4 ]]; then
+			die "Unable to install newer version of 'bash'!"
+		fi
+		echo "Installed bash."
+	fi
+	echo "bash 4+: available"
+}
 
-	if [[ "$EXPECTED_CHECKSUM" != "$ACTUAL_CHECKSUM" ]]; then
+function do_jq {
+	if ! has_command jq; then
+		echo "jq: not found"
+		echo "Installing jq..."
+		brew install jq &>/dev/null
+		if ! has_command jq; then
+			die "Unable to install newer version of jq!"
+		fi
+		echo "Installed jq."
+	fi
+	echo "jq: available"
+}
+
+function do_pnpm {
+	if ! has_command pnpm; then
+		echo "pnpm: not found"
+		echo "Installing pnpm..."
+		curl -fsSL https://get.pnpm.io/install.sh | sh - &>/dev/null
+
+		# Make sure pnpm is in one's path
+		export PNPM_HOME="$("$SHELL" -i -c 'echo "$PNPM_HOME"')"
+		export PATH="$("$SHELL" -i -c 'echo "$PATH"')"
+		if ! has_command pnpm; then
+			die "Unable to install 'pnpm'!"
+		fi
+		echo "Installed pnpm."
+	fi
+	echo "pnpm: available"
+}
+
+function do_php {
+	if ! has_command php || [[ $(php -r "echo version_compare(PHP_VERSION,'$PHP_VERSION');") -eq -1 ]]; then
+		echo "PHP $PHP_VERSION: not found"
+
+		# Note that we can't use grep -v, as it's prematurely terminated.
```

Comment:
The comment had a typo: cc42fbc3ff

I meant to say `grep -q`. See also:
https://stackoverflow.com/questions/67681070/zsh-brew-list-any-command-raises-uncaught-signal-pipe-error

Reply:
Ah. Is brew printing that error message, or zsh? We do similar things in other bash scripts in various places with no problem (or with just a `|| true` to fix the exit code for `-eo pipefail`).

Reply:
It's `brew` that prints it:
https://github.com/Homebrew/brew/blob/821f67ff6db8d4a5b08010d065166faa979e72bf/Library/Homebrew/exceptions.rb#L670-L678

As a more general note: I'm not sure how a `|| true` would help with `grep -q`, as we'd want a non-zero exit code from `grep -q` if not found. 🤔

Reply:
> As a more general note: I'm not sure how a || true would help with grep -q, as we'd want a non-zero exit code from grep -q if not found. 🤔

I was thinking of cases like `VAR=$( ... | grep 'something' )`, where if the `...` bit SIGPIPEs then `-eo pipefail` would make the script exit.

OTOH, even then `{ ... || true; } | grep -q` could be a trick to use so we only get the exit code from `grep` and not `...` (pipefail or otherwise).